### DISCUSSION TITLE
[请教一下] 选的AsIs，为何个别域名命中 geosite:cn，它的子域名还是会对远程和直连 DNS 同时发起解析请求？

### DISCUSSION THREAD

**Discussion by Deleted-Account** [Openwrt-Passwall/openwrt-passwall2#471] [Q&A] [answered]
**为何个别域名命中 geosite:cn，它的子域名还是会对远程和直连 DNS 同时发起解析请求？**

x86 OpenWRT，Passwall2 1.25-1，Xray-core 1.8.7

防火墙有拦截 53 查询的设置
```
iptables -t nat -A PREROUTING -p udp --dport 53 -j REDIRECT --to-ports 53
iptables -t nat -A PREROUTING -p tcp --dport 53 -j REDIRECT --to-ports 53
```
路由器没开 IPv6

我本地架设了两个AdGuardHome，
一个 AGH 上游全是国内 DNS，填入 OpenWRT dnsmasq 的 DNS forwarding，给 Passwall2 的直连调用
一个 AGH 上游全是国外 DOH，填入 OpenWRT 的 remote DNS，Remote DNS outbound 选 Direct，给 Passwall2 remote DNS 调用
![image](https://github.com/xiaorouji/openwrt-passwall2/assets/48331538/aee12bb5-0276-4096-940e-d34b4f2f0eab)

```
2024-01-31 10:00:14: 127.0.0.1#15353 （直连DNS：127.0.0.1#6053 远程DNS：127.0.0.1#7913）
2024-01-31 10:00:15:   - [0]节点列表中的域名(vpslist)：127.0.0.1#30053
2024-01-31 10:00:15:   - [0]默认：127.0.0.1#15353
2024-01-31 10:00:15: 开始加载防火墙规则...
2024-01-31 10:00:15: 加入所有节点到ipset[passwall2_vpslist]直连完成
2024-01-31 10:00:16: 加载路由器自身 TCP 代理...
2024-01-31 10:00:16:   - 启用 TPROXY 模式
2024-01-31 10:00:16: 加载路由器自身 UDP 代理...
2024-01-31 10:00:16: 加入负载均衡的节点到ipset[passwall2_vpslist]直连完成
2024-01-31 10:00:16: TCP默认代理：使用节点[分流总节点] [全局代理](TPROXY:1041)代理所有端口
2024-01-31 10:00:16: UDP默认代理：使用节点[分流总节点] [全局代理](TPROXY:1041)代理所有端口
2024-01-31 10:00:16: 防火墙规则加载完成！
2024-01-31 10:00:19: 重启 dnsmasq 服务
2024-01-31 10:00:19: 配置定时任务：自动更新【MrBlank】订阅。
2024-01-31 10:00:19: 运行完成！
```

Xray 的 Domain Strategy 选的 AsIs  
Domain matcher 选的 hybrid  

规则就是 PW2 默认的顺序，默认节点选直连
![image](https://github.com/xiaorouji/openwrt-passwall2/assets/48331538/519850d1-2796-40b1-ae07-3fa6ec543741)

我阅读了 Xray 的官方文档 (https://xtls.github.io)
以及 Loyalsoldier 的 v2ray-rules-dat 的说明 (https://github.com/Loyalsoldier/v2ray-rules-dat)
还查看了 Passwall2 生成的 xray 配置: /tmp/etc/passwall2/global.json

还是很疑惑啊，为何个别域名命中 geosite:cn 后，子域名还是向远程和直连同时发起解析请求？
![image](https://github.com/xiaorouji/openwrt-passwall2/assets/48331538/5b159034-5406-46d3-8543-6ef55a884411)
比如这个 api.m.jd.com
jd.com 肯定命中 geosite:cn 呀
但是为何还向直连和远程 DNS 同时发起请求？
这期间并没有操作 PW2 或者有 wan 重播 导致 PW2 重启的事件。
**开始我怀疑是因为请求来自 docker，但是查看了下 DNS 的日志，并不是啊，是来自某个台式机**
(我其实是3个 AGH，还有个是host模式使用53端口，此 AGH 上游填 dnsmasq 修改的端口。这个 AGH 没有开 Optimistic caching)

大部域名的子域名又从来没有这种情况
比如 meituan.com、xiaomi.com 等，分流是正常的
类似下图
![image](https://github.com/xiaorouji/openwrt-passwall2/assets/48331538/32e4ebe5-1779-47d5-86c2-d0ea14952285)

**Comment by handsomeji** [comment]
客戶端內置了HTTPS請求。
請在路由器DNS禁用 type65 類型的請求。

**Reply by 2eronia** [reply]
看了query log 排查了下，有点奇怪啊，不是 app 请求的查询出现直连和远程DNS同时查询，请求是来自台式机，Win10。  

涉及到 jd.com 的只有 Edge 等浏览器和相关拓展，Edge 的 secure DNS 是灰色的，没有打开。也没有装什么其他 jd.com 相关的软件了


目前发现子域名直连DNS同时查询的，除了 api.m.jd.com，还有个 hm.baidu.com，
请求来自于一台 Win10 笔记本

**Reply by handsomeji** [reply, marked as answer]
Xray的內置DNS只能處理A/AAAA類型的DNS查詢，其它類型的查詢會轉發到它們原本的目標地址。passwall2是把其他類型的DNS轉發到遠程DNS去解析了。但我突然覺得有必要drop掉，否則可能會影響使用。
Sing-Box的內置DNS應該是支持所有DNS查詢類型的，你可以更換Sing-Box使用一段時間看看。

**Reply by 2eronia** [reply]
感谢您的回复。  

我简单将分流总节点改成了 Sing-box，负载均衡的 socks 也换成了 Sing-box。  
写了个简单的 wget 脚本下载/更新 geosite.db geoip.db 到 /usr/share/singbox 目录 (这步是否必须？印象中在其他帖子有看到要自己下载这两个文件)

跑起来比我预想的轻松，试了下 京东、油管、Twitch……(至少目前) 也如我设想的正常 DNS 分流了

**Reply by xialaup** [reply]
> Xray的內置DNS只能處理A/AAAA類型的DNS查詢，其它類型的查詢會轉發到它們原本的目標地址。passwall2是把其他類型的DNS轉發到遠程DNS去解析了。但我突然覺得有必要drop掉，否則可能會影響使用。
> Sing-Box的內置DNS應該是支持所有DNS查詢類型的，你可以更換Sing-Box使用一段時間看看。

請問啟用singbox分流總節點是不是和這個域名解析策略無關了？而且也沒有在後臺看到singbox的相關設置，祇有xray總節點裡面有喔。

**Reply by xialaup** [reply]
> 感谢您的回复。  
> 
> 我简单将分流总节点改成了 Sing-box，负载均衡的 socks 也换成了 Sing-box。  
> 写了个简单的 wget 脚本下载/更新 geosite.db geoip.db 到 /usr/share/singbox 目录 (这步是否必须？印象中在其他帖子有看到要自己下载这两个文件)
> 
> 跑起来比我预想的轻松，试了下 京东、油管、Twitch……(至少目前) 也如我设想的正常 DNS 分流了

你可以试试这个规则
https://github.com/lyc8503/sing-box-rules


~~`0 5 * * 1 curl -L -o /usr/share/singbox/geoip.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geoip.db
0 5 * * 1 curl -L -o /usr/share/singbox/geosite.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geosite.db
`~~

**Reply by 2eronia** [reply]
大佬您好。更换成 sing-box 内核后，非 A/AAAA 的 DNS 请求也正常分流了。另外有看到您 1.26 的相关 commits


目前遇到的问题是，
计划任务 /sbin/ifup wan 或者 ISP 把我提下线回收 IP 后，Passwall2 会进行一次自动重启，然后 DNS 分流就异常了。
我已更新了相关的 issue
https://github.com/xiaorouji/openwrt-passwall2/issues/470#issuecomment-1955722774

刚才更新了 1.26-1 版本的 Passwall2 和 1.8.5 的 sing-box，等下次 /sbin/ifup wan 或者 ISP 回收 IP 看看情况如何。

目前已知通过
`service passwall2 restart`
重启 Passwall2，DNS 分流是正常的。不正常仅出现在 ISP 回收 IP 导致 IP 变动 或者 /sbin/ifup wan 重播后导致的那次 Passwall2 重启

**Comment by laaArmy** [comment]
困扰我很久的问题

**Comment by 2eronia** [comment]
> > 感谢您的回复。
> > 我简单将分流总节点改成了 Sing-box，负载均衡的 socks 也换成了 Sing-box。
> > 写了个简单的 wget 脚本下载/更新 geosite.db geoip.db 到 /usr/share/singbox 目录 (这步是否必须？印象中在其他帖子有看到要自己下载这两个文件)
> > 跑起来比我预想的轻松，试了下 京东、油管、Twitch……(至少目前) 也如我设想的正常 DNS 分流了
> 
> 你可以试试这个规则 https://github.com/lyc8503/sing-box-rules
> 
> `0 5 * * 1 curl -L -o /usr/share/singbox/geoip.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geoip.db 0 5 * * 1 curl -L -o /usr/share/singbox/geosite.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geosite.db `



> > 感谢您的回复。
> > 我简单将分流总节点改成了 Sing-box，负载均衡的 socks 也换成了 Sing-box。
> > 写了个简单的 wget 脚本下载/更新 geosite.db geoip.db 到 /usr/share/singbox 目录 (这步是否必须？印象中在其他帖子有看到要自己下载这两个文件)
> > 跑起来比我预想的轻松，试了下 京东、油管、Twitch……(至少目前) 也如我设想的正常 DNS 分流了
> 
> 你可以试试这个规则 https://github.com/lyc8503/sing-box-rules
> 
> `0 5 * * 1 curl -L -o /usr/share/singbox/geoip.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geoip.db 0 5 * * 1 curl -L -o /usr/share/singbox/geosite.db https://github.com/lyc8503/sing-box-rules/releases/latest/download/geosite.db `

谢谢。我试试
curl 下载失败是否会变成 0kb 的文件？
我之前写的小脚本用的 wget，因为网络原因更错好几次了，后来加了个判断，更新出错就还原旧版

**Reply by xialaup** [reply]
如果curl不能用就用wget呗
